server/sonar-server/src/test/java/org/sonar/server/edition/ws/ApplyLicenseActionTest.java

```diff
@@ -158,6 +162,54 @@ public void request_fails_with_BadRequestException_if_license_is_invalid() {
   }
 
   @Test
+  public void request_fails_with_ISE_if_is_cluster_and_license_plugin_is_not_installed() throws IOException {
+    underTest = new ApplyLicenseAction(userSessionRule, mutableEditionManagementState, editionInstaller, webServer, null);
+    actionTester = new WsActionTester(underTest);
+    userSessionRule.logIn().setSystemAdministrator();
+
+    when(mutableEditionManagementState.getPendingInstallationStatus()).thenReturn(NONE);
+    when(webServer.isStandalone()).thenReturn(false);
+
+    TestRequest request = actionTester.newRequest().setParam(PARAM_LICENSE, createLicenseParam("dev", "plugin1"));
+
+    expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage("License-manager plugin should be installed");
+    request.execute();
+  }
+
+  @Test
+  public void request_fails_with_ISE_if_no_change_needed_but_license_plugin_is_not_installed() throws IOException {
+    underTest = new ApplyLicenseAction(userSessionRule, mutableEditionManagementState, editionInstaller, webServer, null);
+    actionTester = new WsActionTester(underTest);
+    userSessionRule.logIn().setSystemAdministrator();
+
+    setPendingLicense(NONE);
+    when(webServer.isStandalone()).thenReturn(true);
+
+    TestRequest request = actionTester.newRequest().setParam(PARAM_LICENSE, createLicenseParam("dev", "plugin1"));
+
+    expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage("Can't decide edition does not require install if LicenseCommit instance is null");
+    request.execute();
+  }
+
+  @Test
+  public void always_apply_license_when_is_cluster() throws IOException {
```

Comment:
I have to remember to add a test for following use case once my branch is based on this one: apply license in cluster mode and license update fails (ie. license validation fails => throws IllegalArgumentException)